### PR TITLE
Dumb down some incomprehensible code

### PR DESCRIPTION
**Pull request by lucasart**
Code like this is more a case of showing off one's C++ knowledge, rather than
using it adequately, IMHO.

**First loop (std::generate)**

Iterators are inadequate here, because they lose the key information which is
idx. As a result, we need to carry a redundant idx variable, and increment it
along the way. Very clumsy.
Usage of std::generate and a lambda function only obfuscate the code, which is
merely a simple and stupid loop over the elements of a vector.

**Second loop (std::accumulate)**

This code is thoroughlly incomprehensible. Restore the original, which was much
simpler to understand.

**Third loop (range based loop)**

Again, a range based loop is inadequate, because we lose idx! To resolve this
artificially created problem, the data model was made redundant (idx is a data
member of db[] elements!?), which is ugly and unjustified. A simple and stupid
for loop with idx does the job much better.

No functional change.

### CODE REVIEW

**Comment by mcostalba** [comment]
Because I am the author of the "thoroughly incomprehensible" code, I am not particularly happy with this pull request, but I understand your points and it is ok for me if Joona and Gary decide to merge this one.

**Comment by lucasart** [comment]
@mcostalba: you honestly think that this code

    while (std::accumulate(db.begin(), db.end(), false, [&](bool repeat, KPKPosition& pos)
    { return (pos == UNKNOWN && pos.classify(db) != UNKNOWN) || repeat; })){}

is easier to understand than the original version:

    while (repeat)
        for (repeat = idx = 0; idx < MAX_INDEX; ++idx)
            repeat |= (db[idx] == UNKNOWN && db[idx].classify(db) != UNKNOWN);

?

**Comment by mcostalba** [comment]
As I have written : I understand your points.

**Comment by kalnov** [comment]
This is highly a matter of opinion, but for me, Marco's code is much easier to understand. And it is presumably lucid enough for anybody to understand. There is no absolutely need to dumb it down any further.

**Comment by Nazine** [comment]
im relatively new to c++ and the default code is quite easy to understand. lucas' code is slightly complex

**Comment by DU-jdto** [comment]
Nice sockpuppet, Nazine.

**Comment by zamar** [comment]
Either way, I can read the code without much diificulty. 

It's a question of taste and personally I've never been too interested about coding style discussions/fights...

**Comment by billwritescode** [comment]
I agree with Nazine. Loops are classic and not confusing. I'm new to c++.

**Comment by zamar** [comment]
After a second look, I somewhat favour Lucas' solution... The use of std::accumulate is indeed a bit cryptic.

So, I'll commit the change in the near future unless someone objects.

**Comment by glinscott** [comment]
I'd be fine with the code either way here.  Decision is up to you :).

**Comment by Altorffer** [comment]
Given the fact the pull request has already been merged, the following point might be redundant but, has anyone profiled either approach? Given the fact they are in fact 2 very different approaches, at least visually, there is the possibility the compiler might in fact also end up generating 2 different approaches.

**Comment by DU-jdto** [comment]
@Altorffer The code in question is run once at the beginning of the program. Absurd extremes aside, the speed doesn't really matter.

**Comment by Altorffer** [comment]
Ah. Thanks for clarifying that.
